File: src/nif.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
use crate::{
    cleanup_container, connect_to_docker, container_logs, get_image, setup_container,
    wait_container,
};
use bollard::container::StartContainerOptions;
use futures_util::TryFutureExt;
use once_cell::sync::Lazy;
use std::{env, path::Path};
use tokio::runtime::{Builder, Runtime};

static TOKIO: Lazy<Runtime> = Lazy::new(|| {
    Builder::new_current_thread()
        .enable_all()
        .build()
        .expect("Failed to start tokio runtime")
});

#[rustler::nif]
fn prepare_container() -> String {
    let project_path = Path::new(env!("CARGO_MANIFEST_DIR"));

    let container_name = "funless-node-container";
    let image_name = "node:lts-alpine";
    let tar_file = project_path.join("js/hello.tar.gz").display().to_string();
    let main_file = "/opt/index.js";

    let docker = connect_to_docker("/run/user/1001/docker.sock")
        .expect("Failed to connect to docker socket");
    let f = get_image(&docker, "node:lts-alpine").and_then(|_| {
        setup_container(&docker, &container_name, &image_name, &main_file, &tar_file)
    });

    let result = TOKIO.block_on(f);

    match result {
        Ok(()) => "ok".to_string(),
        Err(e) => e.to_string(),
    }
}

#[rustler::nif]
fn run_function() -> String {
    let container_name = "funless-node-container";

    let docker = connect_to_docker("/run/user/1001/docker.sock")
        .expect("Failed to connect to docker socket");

    let f = docker
        .start_container(container_name, None::<StartContainerOptions<String>>)
        .and_then(|_| wait_container(&docker, &container_name))
        .and_then(|_| container_logs(&docker, &container_name));

    let logs = TOKIO.block_on(f);

    match logs {
        Ok(v) => v[0].to_string(),
        Err(e) => e.to_string(),
    }
}

#[rustler::nif]
fn cleanup() -> String {
    let container_name = "funless-node-container";

    let docker = connect_to_docker("/run/user/1001/docker.sock")
        .expect("Failed to connect to docker socket");

    let result = TOKIO.block_on(cleanup_container(&docker, container_name));

    match result {
        Ok(()) => "ok".to_string(),
        Err(e) => e.to_string(),
    }
}

File: lib/worker/adapters/runtime/openwhisk/nif.ex
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
defmodule Worker.Adapters.Runtime.OpenWhisk.Nif do
  @moduledoc """
  NIFs used to manipulate docker OpenWhisk runtimes.
  """
  use Rustler, otp_app: :worker, crate: :fn_docker

  #   Creates the `_runtime_name` container, with information taken from `_function`.
  #   ## Parameters
  #     - _function: Worker.Domain.Function struct, containing function information
  #     - _runtime_name: name of the container that will be created
  #     - _network_name: name of the network to which the container will be attached
  #     - _docker_host: path of the docker socket in the current system
  @doc false
  def prepare_runtime(_function, _runtime_name, _network_name, _docker_host) do
    :erlang.nif_error(:nif_not_loaded)
  end

  #   Gets the logs of the `_runtime_name` container.
  #   ## Parameters
  #     - _runtime_name: name of the container
  #     - _docker_host: path of the docker socket in the current system
  @doc false
  def runtime_logs(_runtime_name, _docker_host) do
    :erlang.nif_error(:nif_not_loaded)
  end

  #   Removes the `_runtime_name` container.
  #   ## Parameters
  #     - _runtime_name: name of the container that will be removed
  #     - _docker_host: path of the docker socket in the current system
  @doc false
  def cleanup_runtime(_runtime_name, _docker_host) do
    :erlang.nif_error(:nif_not_loaded)
  end
end
